plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -0,0 +1,267 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <gio/gio.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <gudev/gudev.h>
+
+#include "fu-plugin-vfuncs.h"
+#include "fu-device-metadata.h"
+
+#ifndef HAVE_GUDEV_232
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevDevice, g_object_unref)
+#endif
+
+#define TBT_NEW_DEVICE_TIMEOUT	2
+
+struct FuPluginData {
+	GUdevClient   *udev;
+	gboolean       needs_forcepower;
+	gboolean       can_forcepower;
+};
+
+static gchar *
+fu_plugin_thunderbolt_power_get_path (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GList) devices;
+	const gchar *basepath;
+	const gchar *driver;
+	g_autofree gchar *built_path = NULL;
+
+	devices = g_udev_client_query_by_subsystem (data->udev, "wmi");
+	for (GList* l = devices; l != NULL; l = l->next) {
+		g_autoptr(GUdevDevice) device = l->data;
+
+		/* only supports intel-wmi-thunderbolt for now */
+		driver = g_udev_device_get_driver (device);
+		if (g_strcmp0 (driver, "intel-wmi-thunderbolt") != 0)
+			continue;
+
+		/* check for the attribute to be loaded */
+		basepath = g_udev_device_get_sysfs_path (device);
+		if (basepath == NULL)
+			continue;
+		built_path = g_build_path ("/", basepath,
+					   "force_power", NULL);
+		if (g_file_test (built_path, G_FILE_TEST_IS_REGULAR))
+			break;
+		built_path = NULL;
+	}
+
+	return g_steal_pointer (&built_path);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_supported (FuPlugin *plugin)
+{
+	g_autofree gchar *path = NULL;
+	path = fu_plugin_thunderbolt_power_get_path (plugin);
+	if (path != NULL) {
+		g_debug ("Detected force power support at %s", path);
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_set (FuPlugin *plugin, gboolean enable,
+				       GError **error)
+{
+	g_autofree gchar *path = NULL;
+	gint fd;
+	gint ret;
+
+	path = fu_plugin_thunderbolt_power_get_path (plugin);
+	if (path == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "unable to set power to %d (missing kernel support)",
+			     enable);
+		return FALSE;
+	}
+	g_debug ("Setting force power to %d", enable);
+	fd = g_open (path, O_WRONLY);
+	if (fd == -1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "failed to open %s", path);
+		return FALSE;
+	}
+        ret = write (fd, enable ? "1" : "0", 1);
```

Comment:
spaces?

Reply:
Ah that's what I get for copying and pasting from you guys.  My editor didn't fixup, I'll fix in another commit.